### PR TITLE
Switch to Ninja build

### PR DESCRIPTION
**Pull request by ikitayama**
This is build test on X86.

### CODE REVIEW

**Comment by ikitayama** [comment]
@trws @naromero77 please review. If you'd like to stick with Unix generator, that's fine by me, but please consider adding a check in the pre_build() function.

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -581,12 +589,13 @@ def cmake_args(self):
...
         with working_dir(self.build_directory):
-            # When building shared libraries these need to be installed first
-            make("install-LLVMTableGen")
-            if self.spec.version >= Version("4.0.0"):
-                # LLVMDemangle target was added in 4.0.0
-                make("install-LLVMDemangle")
-            make("install-LLVMSupport")
+            if "+shared_libs" in self.spec:
+               # When building shared libraries these need to be installed first
+               make("install-LLVMTableGen")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/llvm/package.py:594]: So, what was the story with this? Did you try `+shared_libs` with the llvm+ninja build? You mentioned that failed https://github.com/spack/spack/issues/23880#issuecomment-848257616

Comment (ikitayama) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: `shared_libs` defaults to `False` so I added the check.

Comment (haampie) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: Yes, but if you have it enabled, does it build with ninja or not?

Comment (haampie) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: Hm, maybe this is outside the scope of making ninja work. I think it makes sense to remove/rename the `+shared_libs` variant to something obscure, since the average user should use `+llvm_dylib`.

Comment (ikitayama) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: > Yes, but if you have it enabled, does it build with ninja or not?

Clearly not.

Comment (ikitayama) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: Now I am wondering that pre build phase manual `make` commands are required or not.

Comment (ikitayama) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: For `llvm` switching to `Ninja` only is a sensible option as it is much faster and the community uses it.

Comment (ikitayama) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: But I'd like to hear @naromero77 and @trws 's opinions on this. What's your take?

Comment (ikitayama) [reply var/spack/repos/builtin/packages/llvm/package.py:594]:
```
       if "+shared_libs" in spec:
            cmake_args.append("-DBUILD_SHARED_LIBS:Bool=ON")
        if "+llvm_dylib" in spec:
            cmake_args.append("-DLLVM_BUILD_LLVM_DYLIB:Bool=ON")
```

I'm stumbled upon these lines.

Comment (haampie) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: Yeah, I noticed these too. I would pressume clang is not really usable with `BUILD_SHARED_LIBS=On`. Should we consider dropping that option @naromero77 / @trws? "Advanced" users can just run `spack build-env llvm+clang` and then configure with -DBUILD_SHARED_LIBS=On by hand?

It seems we are currently already depending on internals of the generated makefile to support BUILD_SHARED_LIBS (iiuc?), and they don't work with ninja.

Edit: @ikitayama I somehow thought this patch supported both with and without ninja, but it does not. Since ninja does not use make files, it doesn't make sense to call `make install ...` at all, so you'll have to remove this `pre_install` section. Can you please check WHY it was added and see if the original problem it solved is not present with the ninja build?

**Review by haampie — CHANGES_REQUESTED** [review]
Apart from fixing the flake8 style issues, the changes to `pre_install` do not make sense; all references to `make` should be removed, not made conditional under an unrelated variant value `if "+shared_libs" in self.spec`. Also it should be figured out what made those `make install-x` statements necessary in the first place, and whether workarounds are still necessary after switching to ninja.

**Comment by alalazo** [comment]
@naromero77

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 1 above)
Comment (naromero77) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: > Now I am wondering that pre build phase manual `make` commands are required or not.

@trws would know more than me

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 2]
```diff
@@ -66,6 +67,11 @@ class Llvm(CMakePackage, CudaPackage):
         default=True,
         description="Build the LLVM C/C++/Objective-C compiler frontend",
     )
+    variant(
+        "clang-tools-extra",
+        default=False,
+        description="Clang Tools",
+    )
```
Comment (naromero77) [inline var/spack/repos/builtin/packages/llvm/package.py:74]: Why is a new variant needed? What does this have to do with switching to Ninja?

**Comment by naromero77** [comment]
> @trws @naromero77 please review. If you'd like to stick with Unix generator, that's fine by me, but please consider adding a check in the pre_build() function.

We cannot switch to ninja if it breaks the variants in this package. I tried switching to Ninja a week or two ago and the changes where more involved than what I first thought so I abandoned it due to lack of time.

Most of the LLVM buildbots run with Ninja. I think it might be faster for people with NFS mounted file systems, but I think most of us won't see much of an improvement in build time.

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 1 above)
Comment (naromero77) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: > For `llvm` switching to `Ninja` only is a sensible option as it is much faster and the community uses it.

I think Ninja-only is a sensible option. But see my comments later on in this review.

**Comment by naromero77** [comment]
To summarize, I had talked to @trws about this over Slack before the U.S. Memorial Day weekend break. In general, we are supportive of the switch to Ninja but neither of us have the time to work on this. So, we can do the switch, if the code is made available via a PR. It will need to be well-tested before we accept it since the impact will be broad.

**Comment by ikitayama** [comment]
This PR is really dedicated to the switch. What kind of tests do you expect?

**Comment by naromero77** [comment]
At minimum, it should not break for x86 and OpenMP offload to NVidia GPUs.

@trws may be able to think of others.

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 1 above)
Comment (trws) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: Ninja only seems fine, but anything in there that used to build with the make-based version should build with the ninja-based version.  Replacing those `which`-based make lines with something from `which('ninja')` should just work, but needs to be tested.

Comment (trws) [reply var/spack/repos/builtin/packages/llvm/package.py:594]: Alternately, it could be re-worked to use cmake instead to make it portable to keep both with something like `cmake("--build", ".", "--target", "install-LLVMTableGen")`.

**Comment by trws** [comment]
The `shared-libs` target needs to build for an otherwise default build as well, at least for now.  Anywhere `make()` appears in the file needs to be replaced and tested, basically.

**Comment by ikitayama** [comment]
Do you build llvm with the shared libs option as an expert?

**Comment by trws** [comment]
For build and patch testing? Yes, yes I do.  Would I ever deploy it in production? No, no I would not.

More to the point, I'm relatively sure this patch would currently break building clang with libomptarget for the cuda backend.  It currently uses make unconditionally to do the rebuild of the target-specific binary objects in the post_install hook.  It may also require TableGen for the llvm_dylib target, that option didn't exist when the pre-build step was added and has never been tested without it.

**Comment by ikitayama** [comment]
Alright. I will prepare a preparatory PR for Ninja to work.

**Comment by bernhardkaindl** [comment]
@ikitayama: #27521 has switched LLVM to Ninja recently, is this PR obsolete now?

**Comment by ikitayama** [comment]
Yes, it is then.
